### PR TITLE
feat: propage sort down pipelines

### LINKED ISSUE #1363: Transform row ordering guarantees
Just learned about this project from [HN](https://news.ycombinator.com/item?id=34181319), it looks really great.

While playing with some of the examples, I noticed that the row ordering guarantees of transforms are unclear to me. In particular I don't understand which transforms (if any?) are guaranteed to leave the order of rows unchanged.

For example, does the query below guarantee that the output rows are ordered by salary?

```
from employees
sort salary
take 3
filter salary > 0
```

What about this one?

```
table x = (
  from employees
  sort salary
)
from x
```

From the generated SQL it appears that the answer is NO in both cases (selecting from a CTE has no ordering guarantees in theory). This implies that the following two queries may also produce different row ordering results:

```
from employees
sort salary
filter salary > 0
```

```
from employees
filter salary > 0
sort salary
```

Maybe that's reasonable, but it would certainly feel surprising. Either way, the exact ordering guarantees of transforms should be documented. Apologies if this is already the case and I missed it.

### PR DESCRIPTION
**Pull request by aljazerzen**
Closes #1363 

This PR also includes lots of refactoring that was needed to accomplish what I wanted.

### CODE REVIEW

File: prql-compiler/src/sql/srq/anchor.rs [hunk 1]
```diff
@@ -226,7 +224,7 @@ pub(super) fn anchor_split(
     ctx: &mut AnchorContext,
     preceding: Vec<SqlTransform<TableRef>>,
     atomic: Vec<SqlTransform<TableRef>>,
-) -> (Vec<SqlTransform<TableRef>>, HashMap<CId, CId>) {
+) -> Vec<SqlTransform<TableRef>> {
```
Comment (max-sixty) [inline prql-compiler/src/sql/srq/anchor.rs:227]: (fwiw I do think simplifying signatures is good per se)

File: prql-compiler/src/sql/srq/postprocess.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
+use std::collections::HashMap;
```
Comment (max-sixty) [inline prql-compiler/src/sql/srq/postprocess.rs:1]: Would be good to have a summary of what's done in `postprocess` if possible (but don't want to slow you down, maybe this is the sort of thing I can add as a test of my understand!)

File: prql-compiler/src/sql/srq/anchor.rs (hunk 1 above)
Comment (aljazerzen) [reply prql-compiler/src/sql/srq/anchor.rs:227]: Yeah, because I store cid_redirects in Context. Which is more complex I'd say.

**Comment by snth** [comment]
👏

**Pull request merged**